File: headers_list/src/main/java/HeadersList/HeadersListFactory.java
package HeadersList;

import java.util.ArrayList;
import java.util.List;

public final class HeadersListFactory {

    private HeadersListFactory() {}

    private static <T extends Comparable<T>> HeadersList<T> copy(HeadersList<T> original) {
        HeadersList<T> newInstance = new HeadersList<T>();

        for (HeadersList.SerializableNode<T> nodeData : original) {
            List<T> associatedListCopy = new ArrayList<>(nodeData.associatedList());
            newInstance.add(nodeData.header(), associatedListCopy);
        }

        return newInstance;
    }

    public static <T extends Comparable<T>> HeadersList<T> add(HeadersList<T> source, T header, List<T> list) {
        HeadersList<T> newInstance = copy(source);
        newInstance.add(header, list);
        return newInstance;
    }

    public static <T extends Comparable<T>> HeadersList<T> add(HeadersList<T> source, int index, T header, List<T> list) {
        HeadersList<T> newInstance = copy(source);
        newInstance.add(index, header, list);
        return newInstance;
    }

    public static <T extends Comparable<T>> HeadersList<T> addSorted(HeadersList<T> source, T header, List<T> list) {
        HeadersList<T> newInstance = copy(source);
        newInstance.addSorted(header, list);
        return newInstance;
    }

    public static <T extends Comparable<T>> HeadersList<T> remove(HeadersList<T> source, int index) {
        HeadersList<T> newInstance = copy(source);
        newInstance.remove(index);
        return newInstance;
    }

    public static <T extends Comparable<T>> HeadersList<T> sort(HeadersList<T> source) {
        HeadersList<T> newInstance = copy(source);
        newInstance.sort();
        return newInstance;
    }

    public static <T extends Comparable<T>> HeadersList<T> balance(HeadersList<T> source) {
        HeadersList<T> newInstance = copy(source);
        newInstance.balance();
        return newInstance;
    }
}

File: headers_list/src/main/kotlin/HeadersList/HeadersListFactory.kt
package HeadersList

object HeadersListFactory {

    private fun <T : Comparable<T>> copy(original: HeadersList<T>): HeadersList<T> {
        val newInstance = HeadersList<T>()

        for (nodeData in original) {
            val associatedListCopy = ArrayList(nodeData.associatedList)
            newInstance.add(nodeData.header, associatedListCopy)
        }

        return newInstance
    }

    fun <T : Comparable<T>> add(source: HeadersList<T>, header: T, list: MutableList<T>): HeadersList<T> {
        val newInstance = copy(source)
        newInstance.add(header, list)
        return newInstance
    }

    fun <T : Comparable<T>> add(source: HeadersList<T>, index: Int, header: T, list: MutableList<T>): HeadersList<T> {
        val newInstance = copy(source)
        newInstance.add(index, header, list)
        return newInstance
    }

    fun <T : Comparable<T>> addSorted(source: HeadersList<T>, header: T, list: MutableList<T>): HeadersList<T> {
        val newInstance = copy(source)
        newInstance.addSorted(header, list)
        return newInstance
    }

    fun <T : Comparable<T>> remove(source: HeadersList<T>, index: Int): HeadersList<T> {
        val newInstance = copy(source)
        newInstance.remove(index)
        return newInstance
    }

    fun <T : Comparable<T>> sort(source: HeadersList<T>): HeadersList<T> {
        val newInstance = copy(source)
        newInstance.sort()
        return newInstance
    }

    fun <T : Comparable<T>> balance(source: HeadersList<T>): HeadersList<T> {
        val newInstance = copy(source)
        newInstance.balance()
        return newInstance
    }
}
